src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,740 @@
+
+    icmp6Header = reinterpret_cast<const Ip6::Icmp::Header *>(aBuffer);
+
+    switch (icmp6Header->GetType())
+    {
+    case Ip6::Icmp::Header::kTypeRouterAdvert:
+        HandleRouterAdvertisement(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    case Ip6::Icmp::Header::kTypeRouterSolicit:
+        HandleRouterSolicit(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    default:
+        ExitNow();
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(IsInitialized());
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+    if (aEvents.Contains(kEventThreadNetdataChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            EvaluateRoutingPolicy();
+        }
+    }
+
+exit:
+    return;
+}
+
+uint8_t RoutingManager::EvaluateOmrPrefix(Ip6::Prefix *aNewOmrPrefixes)
+{
+    uint8_t                         newOmrPrefixesNum = 0;
+    NetworkData::Iterator           iterator          = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        if (newOmrPrefixesNum < kMaxOmrPrefixNumber)
+        {
+            aNewOmrPrefixes[newOmrPrefixesNum++] = onMeshPrefixConfig.GetPrefix();
+        }
+        else
+        {
+            otLogInfoBr("EvaluateOmrPrefix: too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+        }
+    }
+
+    if (newOmrPrefixesNum == 0)
+    {
+        if (!IsLocalOmrPrefixAdvertised())
+        {
+            otLogInfoBr("EvaluateOmrPrefix: no valid OMR prefixes found in Thread network");
+            if (PublishLocalOmrPrefix() == OT_ERROR_NONE)
+            {
+                aNewOmrPrefixes[newOmrPrefixesNum++] = mLocalOmrPrefix;
+            }
+        }
+    }
+    else if (newOmrPrefixesNum > 1 || (aNewOmrPrefixes[0] != mLocalOmrPrefix))
+    {
+        if (IsLocalOmrPrefixAdvertised())
+        {
+            otLogInfoBr("EvaluateOmrPrefix: there is already OMR prefix(es) in the Thread network",
+                        mLocalOmrPrefix.ToString().AsCString());
+
+            // TODO: there is a corner case that two Duckhorn BR may unpublish their
+            //       OMR prefix at the same time. But should be rare. If this happens,
+            //       OMR prefix evaluation will be triggered again.
+            UnpublishLocalOmrPrefix();
```

Comment:
need to remove the local OMR prefix from `aNewOmrPrefixes` and deincrease `newOmrPrefixesNum`?

Reply:
fixed. please review.